OpenRA.Mods.Common/Traits/GivesBounty.cs

```diff
@@ -69,13 +82,24 @@ void INotifyKilled.Killed(Actor self, AttackInfo e)
 			if (info.DeathTypes.Count > 0 && !e.Damage.DamageTypes.Overlaps(info.DeathTypes))
 				return;
 
-			var cost = self.GetSellValue();
-
-			// 2 hundreds because of GetMultiplier and info.Percentage.
-			var bounty = cost * GetMultiplier(self) * info.Percentage / 10000;
-
-			if (info.ShowBounty && bounty > 0 && e.Attacker.Owner.IsAlliedWith(self.World.RenderPlayer))
-				e.Attacker.World.AddFrameEndTask(w => w.Add(new FloatingText(self.CenterPosition, e.Attacker.Owner.Color.RGB, FloatingText.FormatCashTick(bounty), 30)));
+			// Divide by 10000 because of GetMultiplier and info.Percentage.
+			var bounty = GetBountyValue(self, info.Percentage);
+
+			if (info.ShowBounty && self.IsInWorld && bounty > 0 && e.Attacker.Owner.IsAlliedWith(self.World.RenderPlayer))
+			{
+				var displayedBounty = bounty;
+				if (cargo != null)
+				{
+					foreach (var a in cargo.Passengers)
```

Comment:
This loop should move inside GetBountyValue, and then the lookup below should use the trait instead of the info so that this behaves properly when there are recursive transports (infantry in an APC in a naval transport).